test/e2e/e2e_suite_test.go

```diff
@@ -403,10 +408,102 @@ func tearDown(bootstrapClusterProvider bootstrap.ClusterProvider, bootstrapClust
+	}
+	if config.HasVariable(capi_e2e.KubernetesVersionUpgradeTo) {
+		resolveKubernetesVersion(config, versions, capi_e2e.KubernetesVersionUpgradeTo)
+	}
+}
+
+func resolveKubernetesVersion(config *clusterctl.E2EConfig, versions semver.Versions, varName string) {
+	v := getLatestSkuForMinor(config.GetVariable(varName), versions)
+	if _, ok := os.LookupEnv(varName); ok {
+		Expect(os.Setenv(varName, v)).To(Succeed())
+	}
+	config.Variables[varName] = v
+}
+
+// getAllCAPIImageSkus returns all skus for the capi offer under the "cncf-upstream" publisher.
+func getAllCAPIImageSkus(ctx context.Context, location string) []string {
+	settings, err := auth.GetSettingsFromEnvironment()
+	Expect(err).NotTo(HaveOccurred())
+	subscriptionID := settings.GetSubscriptionID()
+	authorizer, err := settings.GetAuthorizer()
+	Expect(err).NotTo(HaveOccurred())
+	imagesClient := compute.NewVirtualMachineImagesClient(subscriptionID)
+	imagesClient.Authorizer = authorizer
+
+	Byf("Finding image skus for offer %s/%s in %s", capiImagePublisher, capiOfferName, location)
+
+	res, err := imagesClient.ListSkus(ctx, location, capiImagePublisher, capiOfferName)
+	Expect(err).NotTo(HaveOccurred())
+
+	var skus []string
+	if res.Value != nil {
+		skus = make([]string, len(*res.Value))
+		for i, sku := range *res.Value {
+			// we have to do this to make sure the SKU has existing images
+			// see https://github.com/Azure/azure-cli/issues/20115.
+			res, err := imagesClient.List(ctx, location, capiImagePublisher, capiOfferName, *sku.Name, "", nil, "")
+			Expect(err).NotTo(HaveOccurred())
+			if res.Value != nil && len(*res.Value) > 0 {
+				skus[i] = *sku.Name
+			}
+		}
+	}
+	return skus
+}
+
+// parseImageSkuNames parses SKU names in format "k8s-1dot17dot2-ubuntu-1804" to extract the Kubernetes version.
+// it returns a sorted list of all k8s versions found.
+func parseImageSkuNames(skus []string) semver.Versions {
+	var capiSku = regexp.MustCompile(`^k8s-(0|[1-9][0-9]*)dot(0|[1-9][0-9]*)dot(0|[1-9][0-9]*)-ubuntu.*$`)
+	versions := make(semver.Versions, len(skus))
+	for i, sku := range skus {
+		match := capiSku.FindStringSubmatch(sku)
+		if len(match) != 0 {
+			versions[i] = semver.MustParse(fmt.Sprintf("%s.%s.%s", match[1], match[2], match[3]))
+		}
+	}
+	semver.Sort(versions)
+	return versions
+}
+
+// getLatestSkuForMinor gets the latest available patch version in the provided list of sku versions that corresponds to the provided k8s version.
+func getLatestSkuForMinor(version string, skus semver.Versions) string {
+	isStable, match := validateStableReleaseString(version)
+	if isStable {
+		major, err := strconv.ParseUint(match[1], 10, 64)
+		Expect(err).NotTo(HaveOccurred())
+		minor, err := strconv.ParseUint(match[2], 10, 64)
+		Expect(err).NotTo(HaveOccurred())
+		var maxPatch *uint64
+		for _, sku := range skus {
+			if sku.Major == major && sku.Minor == minor && (maxPatch == nil || sku.Patch >= *maxPatch) {
```

Comment:
nit: since we are already sorting the skus collection, can we not directly pick `skus[i-1]` where `skus[i] = {version.Major, version.Minor+1}`? Or we can sort descending and pick the first match for `{sku.major, sku.minor}`.